src/mongocxx/test_util/client_helpers.cpp

```diff
@@ -300,6 +301,198 @@ bool is_numeric(types::bson_value::view value) {
+                   "}, increment={" + std::to_string(val.view().get_timestamp().increment) + "}";
+        case bsoncxx::v_noabi::type::k_regex:
+            return "regex={" + val.view().get_regex().regex.to_string() + "}, options={" +
+                   val.view().get_regex().options.to_string() + "}";
+        case bsoncxx::v_noabi::type::k_minkey:
+            return "minkey";
+        case bsoncxx::v_noabi::type::k_maxkey:
+            return "maxkey";
+        case bsoncxx::v_noabi::type::k_decimal128:
+            return val.view().get_decimal128().value.to_string();
+        case bsoncxx::v_noabi::type::k_symbol:
+            return val.view().get_symbol().symbol.to_string();
+        case bsoncxx::v_noabi::type::k_dbpointer:
+            return val.view().get_dbpointer().value.to_string();
+    }
+}
+
+template <typename Container>
+auto size(Container c) -> decltype(std::distance(std::begin(c), std::end(c))) {
+    return std::distance(std::begin(c), std::end(c));
+};
+
+template <typename Element>
+type to_type(const Element& type) {
+    auto type_str = type.get_string().value.to_string();
+    if (type_str == "binData")
+        return type::k_binary;
+    if (type_str == "long")
+        return type::k_int64;
+    if (type_str == "int")
+        return type::k_int32;
+    if (type_str == "objectId")
+        return type::k_oid;
+    if (type_str == "object")
+        return type::k_document;
+    if (type_str == "date")
+        return type::k_date;
+    throw std::logic_error{"unrecognized element type '" + type_str + "'"};
+}
+
+bool is_set(types::bson_value::view val) {
+    switch (val.type()) {
+        case type::k_null:
+            return false;
+        case type::k_document:
+            return !val.get_document().value.empty();
+        case type::k_array:
+            return !val.get_array().value.empty();
+        default:
+            return true;
+    }
+}
+
+void assert_special_operator(types::bson_value::view actual, document::view expected) {
+    auto op = *expected.begin();
+    REQUIRE(op.key().starts_with("$$"));  // assert special operator
+
+    if (op.key().to_string() == "$$type") {
+        auto actual_t = actual.type();
+        if (op.type() == bsoncxx::v_noabi::type::k_string) {
+            REQUIRE(actual_t == to_type(op));
+            return;
+        }
+
+        REQUIRE(op.type() == bsoncxx::v_noabi::type::k_array);
+        for (auto t : op.get_array().value) {
+            if (actual_t == to_type(t)) {
+                return;
+            }
+        }
+
+        FAIL("type '" + to_string(actual_t) + "' expected in array '" +
+             to_json(op.get_array().value) + "'");
+    } else if (op.key().to_string() == "$$unsetOrMatches") {
+        auto val = op.get_value();
+        if (is_set(actual))
+            assert_matches(actual, val);
+    } else if (op.key().to_string() == "$$sessionLsid") {
+        // TODO: get sessions IDs from entity map
```

Comment:
This is the only outstanding logic. I have to figure out a way to call back to the entity map to verify this special operation's values and the one below. Trying to pass the map through `assert_matches` doesn't work. The entity map depends on this, so if this also depended on the entity map, that would create a circular dependency.